Enable alarm feature for lptmr

ZhaoxiangJin opened the pull request:
LPTMR is a commonly used wake-up source for many NXP MCUs and may be used as the companion low-power timer for the Cortex-M systick. This commit enabled LPTMR set and cacnel ALARM functions.

Enabling single-shot alarm support requires reprogramming the hardware compare register (COMPARE/CMR), which is the same resource used for the periodic TOP callback. Because of this hardware limitation, we added Kconfig 'CONFIG_COUNTER_MCUX_LPTMR_ALARM' to select whether alarm support is enabled. When alarm support is enabled the driver will not allow support set TOP operation at runtime (attempts to set TOP will return -ENOTSUP); When alarm support is
disabled the driver will not allow set and cancel ALARM operation at runtime (attempts to set and cancel ALARM will return -ENOTSUP). Choose the configuration that matches your application's requirements.

Holt-Sun commented on drivers/counter/counter_mcux_lptmr.c:
```diff
@@ -117,16 +154,141 @@ static void mcux_lptmr_isr(const struct device *dev)
+			LPTMR_StartTimer(config->base);
+		}
+
+		callback(dev, 0, data->alarm_target_ticks, data->alarm_user_data);
```
This logic is wrong: you set data->alarm_target_ticks = 0, set data->alarm_user_data = NULL, then pass them to callback, no meaning. You should firstly store data->alarm_target_ticks and data->alarm_user_data, then clear them, and pass the stored into callback.

Holt-Sun commented on drivers/counter/counter_mcux_lptmr.c:
```diff
@@ -117,16 +154,141 @@ static void mcux_lptmr_isr(const struct device *dev)
+		if (delta > data->current_top) { /* Too large delta */
+			return -EINVAL;
+		}
+
```
```suggestion
        /* Disallow zero period; coerce to one tick for immediate-ish firing */
        if (delta == 0U) {
            delta = 1U;
        }
```

Holt-Sun commented on drivers/counter/counter_mcux_lptmr.c:
```diff
@@ -117,16 +154,135 @@ static void mcux_lptmr_isr(const struct device *dev)
+
+		uint32_t current_count = LPTMR_GetCurrentTimerCount(config->base);
+
+		callback(dev, 0, current_count, user_data);
```
Here is the defintion of 
```c
/** @brief Alarm callback
 *
 * @param dev       Pointer to the device structure for the driver instance.
 * @param chan_id   Channel ID.
 * @param ticks     Counter value that triggered the alarm.
 * @param user_data User data.
 */
typedef void (*counter_alarm_callback_t)(const struct device *dev,
					 uint8_t chan_id, uint32_t ticks,
					 void *user_data);
```
The "ticks" should be alarm_target_ticks instead of current_count.

Holt-Sun commented on drivers/counter/counter_mcux_lptmr.c:
```diff
@@ -117,16 +154,135 @@ static void mcux_lptmr_isr(const struct device *dev)
+static int mcux_lptmr_set_alarm(const struct device *dev, uint8_t chan_id,
+				const struct counter_alarm_cfg *alarm_cfg)
+{
+	ARG_UNUSED(chan_id);
```
Should validate chan_id, only chan_id 0 is supported.

ZhaoxiangJin replied:
counter common API already [validated ](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/counter.h#L505)this

zejiang0jason commented on drivers/counter/counter_mcux_lptmr.c:
```diff
@@ -117,16 +154,139 @@ static void mcux_lptmr_isr(const struct device *dev)
+	 * CNR[COUNTER] is reset when LPTMR is disabled.
+	 */
+	LPTMR_StopTimer(config->base);
+	LPTMR_DisableInterrupts(config->base, kLPTMR_TimerInterruptEnable);
```
Actually, I think support alarm feature by start/stop LPTMR is not a good idea, because it will results in drift, especially when the document has mentioned it:
<img width="827" height="106" alt="image" src="https://github.com/user-attachments/assets/6d1c9ac5-a2e4-4c23-85b8-84682b9a64eb" />

ZhaoxiangJin replied:
It seems that as long as it's in counter mode, regardless of what method is used to implement the alarm, the issue mentioned in the RM NOTE will exist.

ZhaoxiangJin commented:
> I suddenly remember that I have created one PR to add alarm and guard period API functions before: #93302 The PR is closed because "using the same counter register for both top callback and alarms? both APIs can be used simultaneously".

Did you get counter_basic_api testcase passed? testcases `test_set_top_value_with_alarm `and `single_shot_alarm_top ` seems test top and alarm simultaneously.

Holt-Sun commented:
> > I suddenly remember that I have created one PR to add alarm and guard period API functions before: #93302 The PR is closed because "using the same counter register for both top callback and alarms? both APIs can be used simultaneously".
> 
> Did you get counter_basic_api testcase passed? testcases `test_set_top_value_with_alarm `and `single_shot_alarm_top ` seems test top and alarm simultaneously.

I checked the test_set_top_value_with_alarm and single_shot_alarm_top, they test alarm and top in a function but sequential. The counter_top_cfg should be with flag COUNTER_TOP_CFG_DONT_RESET, then the COUNTER_TOP_CFG_DONT_RESET shall be true, otherwise it will reset the alarm counter.

I understand the top and alarm are designed for different counters with the same top.

ZhaoxiangJin commented:
Update:
Considering that enabling both alarm and top (especially top callback) functionality simultaneously in complex usage scenarios can indeed lead to race conditions, I have added a Kconfig option `CONFIG_COUNTER_MCUX_LPTMR_ALARM` to isolate the two functionalities, allowing users to enable the required functionality based on their needs

Hello @Holt-Sun, @zejiang0jason, please help to review again, thanks.

ZhaoxiangJin committed:
drivers: counter: mcux_lptmr: add alarm support

LPTMR is a commonly used wake-up source for many NXP MCUs and
may be used as the companion low-power timer for the Cortex-M
systick. This commit enabled LPTMR set and cancel ALARM functions.

Enabling single-shot alarm support requires reprogramming
the hardware compare register (COMPARE/CMR), which is the same
resource used for the periodic TOP callback. Because of this
hardware limitation, we added Kconfig 'CONFIG_COUNTER_MCUX_LPTMR_ALARM'
to select whether alarm support is enabled. When alarm support is
enabled the driver will not allow support set TOP operation at runtime
(attempts to set TOP will return -ENOTSUP); When alarm support is
disabled the driver will not allow set and cancel ALARM operation
at runtime (attempts to set and cancel ALARM will return -ENOTSUP).
Choose the configuration that matches your application's requirements.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

zejiang0jason commented on drivers/counter/counter_mcux_lptmr.c:
```diff
@@ -122,9 +227,34 @@ static void mcux_lptmr_isr(const struct device *dev)
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+	counter_alarm_callback_t callback = data->alarm_callback;
+
+	if ((callback != NULL) && (data->alarm_active)) {
```
if callback is NULL, but `data->alarm_active` is true, is it not necessary to stop the alarm?
